### PR TITLE
Fix: crash when SQLJob has unknown dependencies

### LINKED ISSUE #577: [Bug] Creating an SQLJob with "unknown" dependency get the operator to crash
**Documentation**
- [x ] I acknowledge that I have read the relevant [documentation](https://github.com/mariadb-operator/mariadb-operator/tree/main/docs).

**Describe the bug**
<!--
If you create an SQLJob with mariadb operator ghcr.io/mariadb-operator/mariadb-operator:v0.0.27 you get it to crash with the following log:
```
{"level":"info","ts":1713790751.308197,"msg":"Dependency 'job-not-created' not found","controller":"sqljob","controllerGroup":"k8s.mariadb.com","controllerKind":"SqlJob","SqlJob":{"name":"with-with-wrong-dependency-058a94b62a","namespace":"bridge"},"namespace":"bridge","name":"with-with-wrong-dependency-058a94b62a","reconcileID":"e5375ccf-9a74-4e17-ad89-7fbec7b6baee"}
{"level":"info","ts":1713790751.3222082,"msg":"Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference","controller":"sqljob","controllerGroup":"k8s.mariadb.com","controllerKind":"SqlJob","SqlJob":{"name":"with-with-wrong-dependency-058a94b62a","namespace":"bridge"},"namespace":"bridge","name":"with-with-wrong-dependency-058a94b62a","reconcileID":"e5375ccf-9a74-4e17-ad89-7fbec7b6baee"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x348 pc=0x1712c46]

goroutine 579 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/internal/controller/controller.go:116 +0x1e5
panic({0x193a800?, 0x2cb8ac0?})
	/usr/local/go/src/runtime/panic.go:914 +0x21f
github.com/mariadb-operator/mariadb-operator/api/v1alpha1.(*SqlJob).IsComplete(...)
	/app/api/v1alpha1/sqljob_types.go:103
github.com/mariadb-operator/mariadb-operator/controller.(*SqlJobReconciler).waitForDependencies(0xc0003f4d70, {0x1e895f0, 0xc000d95cb0}, 0xc000143180)
	/app/controller/sqljob_controller.go:132 +0x2e6
github.com/mariadb-operator/mariadb-operator/controller.(*SqlJobReconciler).Reconcile(0xc0003f4d70, {0x1e895f0, 0xc000d95cb0}, {{{0xc000afd826?, 0x5?}, {0xc000afeea0?, 0xc000cc5d48?}}})
	/app/controller/sqljob_controller.go:61 +0x1c5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1e8d788?, {0x1e895f0?, 0xc000d95cb0?}, {{{0xc000afd826?, 0xb?}, {0xc000afeea0?, 0x0?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/internal/controller/controller.go:119 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000122aa0, {0x1e89628, 0xc000416ff0}, {0x19cddc0?, 0xc000c57b00?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/internal/controller/controller.go:316 +0x3cc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000122aa0, {0x1e89628, 0xc000416ff0})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/internal/controller/controller.go:266 +0x1c9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/internal/controller/controller.go:227 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 132
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.2/pkg/internal/controller/controller.go:223 +0x565```
for code blocks of your kubectl output or YAML files.
-->

**Expected behaviour**
<!--A concise description of what you expected to happen.-->
Not starting the SqlJob before the dependency exists and are OK. Not crashing

**Steps to reproduce the bug**
<!--Steps to reproduce the bug should be clear and easily reproducible to help people
gain an understanding of the problem.-->
Create a job like:

```
apiVersion: k8s.mariadb.com/v1alpha1
kind: SqlJob
metadata:
  name: my-name
spec:
  restartPolicy: Never
  dependsOn:
    - name: not-existing-job
  mariaDbRef:
    name: mariadb
  username: root
  passwordSecretKeyRef:
    name: mariadb-root
    key: password
  database: my-db
  sql: "use db"
```

**Debug information**
I have added logs above, please let me know if you want anything more.

**Environment details**:
- Kubernetes version: 1.28.3
- mariadb-operator version: 0.0.27
- Install method: helm+argocd

**Additional context**
<!--Add any other context  here.-->

### PR DESCRIPTION
**Pull request by am6737**
Added error handling for unresolved dependencies to prevent the operator from crashing.

Fixes #577

### CODE REVIEW

**Comment by am6737** [comment]
Hey @mmontes11

I have some questions, and I would like to hear your thoughts

1. **Should we iterate through all DependsOn items?**  
   Should we return immediately (on the first error) or iterate through all dependencies and return after checking all of them?

2. **Is it necessary to add a retry interval?**  
   Should we add `RequeueAfter`, or use the default retry mechanism?

3. **Should we directly return return false, ctrl.Result{}, nil without any retries?**

**Comment by mmontes11** [comment]
Hey there @am6737 ! 

> Should we iterate through all DependsOn items?

No, we can return whenever we find a non ready dependency.

> Is it necessary to add a retry interval?

Yes, I think it is valuable to let the user tune the retry interval i.e. linear customizable retry interval. Whenever the default exponential retry  is in advanced state, it can seem that the the `SqlJob` does not get reconciled, therefore I think a linear reconciliation customizable by the user is better in this case.

The source of inspiration is Flux's `retryInterval` and `interval` respectively:
- https://fluxcd.io/flux/components/kustomize/api/v1/
- https://fluxcd.io/flux/components/helm/api/v2/

> Should we directly return return false, ctrl.Result{}, nil without any retries?

I think this is not strictly related to the issue. See a fixed version of `waitForDependencies`:

```go
func (r *SqlJobReconciler) waitForDependencies(ctx context.Context, sqlJob *v1alpha1.SqlJob) (bool, ctrl.Result, error) {
	if sqlJob.Spec.DependsOn == nil {
		return true, ctrl.Result{}, nil
	}
	logger := log.FromContext(ctx)

	for _, dep := range sqlJob.Spec.DependsOn {
		sqlJobDep, err := r.RefResolver.SqlJob(ctx, &dep, sqlJob.Namespace)

		if err != nil {
			msg := fmt.Sprintf("Error getting SqlJob dependency: %v", err)
			if apierrors.IsNotFound(err) {
				msg = fmt.Sprintf("Dependency '%s' not found", dep.Name)
			}

			logger.Info(msg)
			return false, ctrl.Result{RequeueAfter: r.RequeueInterval}, r.patchStatus(ctx, sqlJob, r.ConditionComplete.PatcherFailed(msg))
		}
		if !sqlJobDep.IsComplete() {
			msg := fmt.Sprintf("Dependency '%s' not ready", dep.Name)

			logger.Info(msg)
			if err := r.patchStatus(ctx, sqlJob, r.ConditionComplete.PatcherFailed(msg)); err != nil {
				return false, ctrl.Result{}, err
			}
			return false, ctrl.Result{RequeueAfter: r.RequeueInterval}, nil
		}
	}
	return true, ctrl.Result{}, nil
}
``` 



I think it will be great to add an integration tests for this here to ensure that the `SqlJob` consistently reports non ready state whenever a non existing dependency is specified:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/internal/controller/sqljob_controller_test.go

**Review by mmontes11 — CHANGES_REQUESTED** [review]
Hey there @am6737 ! 

> Should we iterate through all DependsOn items?

No, we can return whenever we find a non ready dependency.

> Is it necessary to add a retry interval?

Yes, I think it is valuable to let the user tune the retry interval i.e. linear customizable retry interval. Whenever the default exponential retry  is in advanced state, it can seem that the the `SqlJob` does not get reconciled, therefore I think a linear reconciliation customizable by the user is better in this case.

The source of inspiration is Flux's `retryInterval` and `interval` respectively:
- https://fluxcd.io/flux/components/kustomize/api/v1/
- https://fluxcd.io/flux/components/helm/api/v2/

> Should we directly return return false, ctrl.Result{}, nil without any retries?

I think this is not strictly related to the issue. See a fixed version of `waitForDependencies`:

```go
func (r *SqlJobReconciler) waitForDependencies(ctx context.Context, sqlJob *v1alpha1.SqlJob) (bool, ctrl.Result, error) {
	if sqlJob.Spec.DependsOn == nil {
		return true, ctrl.Result{}, nil
	}
	logger := log.FromContext(ctx)

	for _, dep := range sqlJob.Spec.DependsOn {
		sqlJobDep, err := r.RefResolver.SqlJob(ctx, &dep, sqlJob.Namespace)

		if err != nil {
			msg := fmt.Sprintf("Error getting SqlJob dependency: %v", err)
			if apierrors.IsNotFound(err) {
				msg = fmt.Sprintf("Dependency '%s' not found", dep.Name)
			}

			logger.Info(msg)
			return false, ctrl.Result{RequeueAfter: r.RequeueInterval}, r.patchStatus(ctx, sqlJob, r.ConditionComplete.PatcherFailed(msg))
		}
		if !sqlJobDep.IsComplete() {
			msg := fmt.Sprintf("Dependency '%s' not ready", dep.Name)

			logger.Info(msg)
			if err := r.patchStatus(ctx, sqlJob, r.ConditionComplete.PatcherFailed(msg)); err != nil {
				return false, ctrl.Result{}, err
			}
			return false, ctrl.Result{RequeueAfter: r.RequeueInterval}, nil
		}
	}
	return true, ctrl.Result{}, nil
}
``` 



I think it will be great to add an integration tests for this here to ensure that the `SqlJob` consistently reports non ready state whenever a non existing dependency is specified:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/internal/controller/sqljob_controller_test.go

**Comment by am6737** [comment]
> Hey there @am6737 !
> 
> > Should we iterate through all DependsOn items?
> 
> No, we can return whenever we find a non ready dependency.
> 
> > Is it necessary to add a retry interval?
> 
> Yes, I think it is valuable to let the user tune the retry interval i.e. linear customizable retry interval. Whenever the default exponential retry is in advanced state, it can seem that the the `SqlJob` does not get reconciled, therefore I think a linear reconciliation customizable by the user is better in this case.
> 
> The source of inspiration is Flux's `retryInterval` and `interval` respectively:
> 
> * https://fluxcd.io/flux/components/kustomize/api/v1/
> * https://fluxcd.io/flux/components/helm/api/v2/
> 
> > Should we directly return return false, ctrl.Result{}, nil without any retries?
> 
> I think this is not strictly related to the issue. See a fixed version of `waitForDependencies`:
> 
> ```go
> func (r *SqlJobReconciler) waitForDependencies(ctx context.Context, sqlJob *v1alpha1.SqlJob) (bool, ctrl.Result, error) {
> 	if sqlJob.Spec.DependsOn == nil {
> 		return true, ctrl.Result{}, nil
> 	}
> 	logger := log.FromContext(ctx)
> 
> 	for _, dep := range sqlJob.Spec.DependsOn {
> 		sqlJobDep, err := r.RefResolver.SqlJob(ctx, &dep, sqlJob.Namespace)
> 
> 		if err != nil {
> 			msg := fmt.Sprintf("Error getting SqlJob dependency: %v", err)
> 			if apierrors.IsNotFound(err) {
> 				msg = fmt.Sprintf("Dependency '%s' not found", dep.Name)
> 			}
> 
> 			logger.Info(msg)
> 			return false, ctrl.Result{RequeueAfter: r.RequeueInterval}, r.patchStatus(ctx, sqlJob, r.ConditionComplete.PatcherFailed(msg))
> 		}
> 		if !sqlJobDep.IsComplete() {
> 			msg := fmt.Sprintf("Dependency '%s' not ready", dep.Name)
> 
> 			logger.Info(msg)
> 			if err := r.patchStatus(ctx, sqlJob, r.ConditionComplete.PatcherFailed(msg)); err != nil {
> 				return false, ctrl.Result{}, err
> 			}
> 			return false, ctrl.Result{RequeueAfter: r.RequeueInterval}, nil
> 		}
> 	}
> 	return true, ctrl.Result{}, nil
> }
> ```
> 
> The problem is that whenever the `RefResolver` returns an `err`, `sqlJobDep` will always be nil, leading to a crash when calling `sqlJobDep.IsComplete`. The solution is returning inside the `err != nil` block.
> 
> I think it will be great to add an integration tests for this here to ensure that the `SqlJob` consistently reports non ready state whenever a non existing dependency is specified:
> 
> * https://github.com/mariadb-operator/mariadb-operator/blob/main/internal/controller/sqljob_controller_test.go

Thank you so much for your example, I completely agree with your approach.

As for the integration test you mentioned, I’m still unsure how to write it properly. I attempted to write a test case myself, but I’m not certain if it’s correct. Could you kindly review it and provide some suggestions? I would really appreciate your help! Here’s the test code I’ve written, where I added a new It block under the Describe section:

```go
It("should report unready status for non-existent dependencies", func() {
		nonExistentDependency := "not-existing-job"
		sqlJob := mariadbv1alpha1.SqlJob{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "sqljob-with-non-existent-dependency",
				Namespace: testNamespace,
			},
			Spec: mariadbv1alpha1.SqlJobSpec{
				DependsOn: []mariadbv1alpha1.LocalObjectReference{
					{
						Name: nonExistentDependency,
					},
				},
				MariaDBRef: mariadbv1alpha1.MariaDBRef{
					ObjectReference: mariadbv1alpha1.ObjectReference{
						Name: testMdbkey.Name,
					},
					WaitForIt: true,
				},
				Username: testUser,
				PasswordSecretKeyRef: mariadbv1alpha1.SecretKeySelector{
					LocalObjectReference: mariadbv1alpha1.LocalObjectReference{
						Name: testPwdKey.Name,
					},
					Key: testPwdSecretKey,
				},
				Database: &testDatabase,
			},
		}

		By("Creating SqlJob with non-existent dependency")
		Expect(k8sClient.Create(testCtx, &sqlJob)).To(Succeed())
		DeferCleanup(func() {
			Expect(k8sClient.Delete(testCtx, &sqlJob)).To(Succeed())
		})

		By("Expecting SqlJob to report unready status")
		Eventually(func() bool {
			var updatedSqlJob mariadbv1alpha1.SqlJob
			if err := k8sClient.Get(testCtx, client.ObjectKeyFromObject(&sqlJob), &updatedSqlJob); err != nil {
				return false
			}
			for _, condition := range updatedSqlJob.Status.Conditions {
				if condition.Type == mariadbv1alpha1.ConditionTypeComplete {
					return condition.Status == metav1.ConditionFalse &&
						condition.Reason == mariadbv1alpha1.ConditionReasonFailed &&
						condition.Message == "Dependency 'not-existing-job' not found"
				}
			}
			return false
		}, testHighTimeout, testInterval).Should(BeTrue())

		By("Expecting SqlJob to not create a Job")
		Consistently(func() bool {
			var job batchv1.Job
			err := k8sClient.Get(testCtx, client.ObjectKeyFromObject(&sqlJob), &job)
			return err != nil && apierrors.IsNotFound(err)
		}, testHighTimeout, testInterval).Should(BeTrue())
	})
```

**Review by mmontes11 — CHANGES_REQUESTED** [review]
Hey @am6737 ! Thanks for the changes.

Your testing approach is good, but I suggest we use gomega's [Consistently](https://pkg.go.dev/github.com/onsi/gomega#Consistently) matcher and the `IsComplete` function from the `SqlJob. See:

```go
	It("should report unready status for non-existent dependencies", func() {
		nonExistentDependency := "not-existing-job"
		sqlJob := mariadbv1alpha1.SqlJob{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "sqljob-with-non-existent-dependency",
				Namespace: testNamespace,
			},
			Spec: mariadbv1alpha1.SqlJobSpec{
				DependsOn: []mariadbv1alpha1.LocalObjectReference{
					{
						Name: nonExistentDependency,
					},
				},
				MariaDBRef: mariadbv1alpha1.MariaDBRef{
					ObjectReference: mariadbv1alpha1.ObjectReference{
						Name: testMdbkey.Name,
					},
					WaitForIt: true,
				},
				Username: testUser,
				PasswordSecretKeyRef: mariadbv1alpha1.SecretKeySelector{
					LocalObjectReference: mariadbv1alpha1.LocalObjectReference{
						Name: testPwdKey.Name,
					},
					Key: testPwdSecretKey,
				},
				Database: &testDatabase,
			},
		}

		By("Creating SqlJob with non-existent dependency")
		Expect(k8sClient.Create(testCtx, &sqlJob)).To(Succeed())
		DeferCleanup(func() {
			Expect(k8sClient.Delete(testCtx, &sqlJob)).To(Succeed())
		})

		By("Expecting SqlJob to report unready status consistently")
		Consistently(func() bool {
			var updatedSqlJob mariadbv1alpha1.SqlJob
			if err := k8sClient.Get(testCtx, client.ObjectKeyFromObject(&sqlJob), &updatedSqlJob); err != nil {
				return false
			}
			return !updatedSqlJob.IsComplete()
		}, 10*time.Second, testInterval).Should(BeTrue())

		By("Expecting SqlJob to not create a Job")
		Consistently(func() bool {
			var job batchv1.Job
			err := k8sClient.Get(testCtx, client.ObjectKeyFromObject(&sqlJob), &job)
			return err != nil && apierrors.IsNotFound(err)
		}, testHighTimeout, testInterval).Should(BeTrue())
```

Let me know if this makes sense to you!

**Comment by am6737** [comment]
> Hey @am6737 ! Thanks for the changes.
> 
> Your testing approach is good, but I suggest we use gomega's [Consistently](https://pkg.go.dev/github.com/onsi/gomega#Consistently) matcher and the `IsComplete` function from the `SqlJob`. See:
> 
> ```go
> 	It("should report unready status for non-existent dependencies", func() {
> 		nonExistentDependency := "not-existing-job"
> 		sqlJob := mariadbv1alpha1.SqlJob{
> 			ObjectMeta: metav1.ObjectMeta{
> 				Name:      "sqljob-with-non-existent-dependency",
> 				Namespace: testNamespace,
> 			},
> 			Spec: mariadbv1alpha1.SqlJobSpec{
> 				DependsOn: []mariadbv1alpha1.LocalObjectReference{
> 					{
> 						Name: nonExistentDependency,
> 					},
> 				},
> 				MariaDBRef: mariadbv1alpha1.MariaDBRef{
> 					ObjectReference: mariadbv1alpha1.ObjectReference{
> 						Name: testMdbkey.Name,
> 					},
> 					WaitForIt: true,
> 				},
> 				Username: testUser,
> 				PasswordSecretKeyRef: mariadbv1alpha1.SecretKeySelector{
> 					LocalObjectReference: mariadbv1alpha1.LocalObjectReference{
> 						Name: testPwdKey.Name,
> 					},
> 					Key: testPwdSecretKey,
> 				},
> 				Database: &testDatabase,
> 			},
> 		}
> 
> 		By("Creating SqlJob with non-existent dependency")
> 		Expect(k8sClient.Create(testCtx, &sqlJob)).To(Succeed())
> 		DeferCleanup(func() {
> 			Expect(k8sClient.Delete(testCtx, &sqlJob)).To(Succeed())
> 		})
> 
> 		By("Expecting SqlJob to report unready status consistently")
> 		Consistently(func() bool {
> 			var updatedSqlJob mariadbv1alpha1.SqlJob
> 			if err := k8sClient.Get(testCtx, client.ObjectKeyFromObject(&sqlJob), &updatedSqlJob); err != nil {
> 				return false
> 			}
> 			return !updatedSqlJob.IsComplete()
> 		}, 10*time.Second, testInterval).Should(BeTrue())
> 
> 		By("Expecting SqlJob to not create a Job")
> 		Consistently(func() bool {
> 			var job batchv1.Job
> 			err := k8sClient.Get(testCtx, client.ObjectKeyFromObject(&sqlJob), &job)
> 			return err != nil && apierrors.IsNotFound(err)
> 		}, 10*time.Second, testInterval).Should(BeTrue())
> ```
> 
> Please, note that I have updated the timeout for `Consistenly`, ensuring the state for `10s` should be enough. Otherwise we would be making assertions during `5m` (`testHighTimeout`).
> 
> Let me know if this makes sense to you!

Thanks for the feedback! I'll go ahead and update the test accordingly with the changes you suggested.

Your input is really helpful, I appreciate it!

File: internal/controller/sqljob_controller.go [hunk 1]
```diff
@@ -128,6 +128,7 @@ func (r *SqlJobReconciler) waitForDependencies(ctx context.Context, sqlJob *v1al
 			if err := r.patchStatus(ctx, sqlJob, r.ConditionComplete.PatcherFailed(msg)); err != nil {
 				return false, ctrl.Result{}, err
 			}
+			return false, ctrl.Result{}, err
```
Comment (mmontes11) [inline internal/controller/sqljob_controller.go:131]: This is no longer needed, see the suggested solution here:
https://github.com/mariadb-operator/mariadb-operator/pull/924#pullrequestreview-2362478989

**Review by mmontes11 — CHANGES_REQUESTED** [review]

**Review by mmontes11 — CHANGES_REQUESTED** [review]
LGTM! Please fix the linting and rebase with `main`. Then we can merge,

**Review by mmontes11 — APPROVED** [review]
LGTM! 🚀 Thanks a lot for this contribution @am6737 🙇🏻

**Pull request merged**